blockchain/validators/src/avr_validator.rs

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+//! Types for reading and writing historical Attestation Verification
+//! Reports from MobileCoin consensus enclaves
+
+use hex::ToHex;
+use mc_attest_core::VerificationReport;
+use mc_common::ResponderId;
+use mc_crypto_keys::Ed25519Public;
+use serde::{Deserialize, Serialize};
+use std::{collections::HashMap, string::String, vec::Vec};
+use toml::{map::Map, value::Value};
+
+/// Object representing AVR history for all consensus nodes
+#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
+pub struct AVRRecords {
+    /// AVR Records
+    pub avr_records: HashMap<ResponderId, Vec<AVRRecord>>,
+}
+
+impl AVRRecords {
+    /// Create new records object
+    pub fn new(avr_records: HashMap<ResponderId, Vec<AVRRecord>>) -> Self {
+        Self { avr_records }
+    }
+}
+
+/// Stores a single historical AVR for each tuple of responder_id
+/// & block signing key  in the format specified in draft MCIP #43
+#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
+pub struct AVRRecord {
+    /// Uri of the consensus node
+    pub responder_id: ResponderId,
+
+    /// Block signing key of the consensus enclave
+    pub block_signing_key: Option<Ed25519Public>,
+
+    /// Block the AVR for the signing key becomes valid
+    pub first_block_index: u64,
+
+    /// Final block the AVR for the signing key is valid
+    pub last_block_index: u64,
+
+    /// Hex of the AVR bytes
+    pub avr: Option<VerificationReport>,
```

Comment:
It really doesn't make sense to have an AVR record without an AVR...

Reply:
Missing data/reports is a valid case that exists in the actual data. There are in fact gaps in the AVRs. This could also be treated by not having any records, but I think that leads to a less explicit handling when later doing backfilling or verification.

Reply:
The intermediate serialization trick has solved this!